Mainnet/HashBattle/HashBattle/Arena.cs

```diff
@@ -0,0 +1,271 @@
+
+        Owner = pendingOwner;
+        PendingOwner = Address.Zero;
+
+        Log(new ClaimPendingDeployerOwnershipLog { From = oldOwner, To = pendingOwner });
+    }
+    /// <summary>
+    /// Battle owner will start the battle
+    /// </summary>
+    public ulong StartBattle(ulong fee)
+    {
+        Assert(Message.Sender == Owner, "Only battle owner can start game.");
+        Assert(fee < ulong.MaxValue / MaxUsers, "Fee is too high");
+
+        var battleId = NextBattleId;
+        NextBattleId += 1;
+
+        var battle = new BattleMain
+        {
+            BattleId = battleId,
+            Fee = fee,
+            Users = new Address[MaxUsers]
+        };
+        SetBattle(battleId, battle);
+
+        Log(new BattleEventLog { Event = "Start", BattleId = battleId, Address = Message.Sender });
+        return battleId;
+    }
+    /// <summary>
+    /// 4 different user will enter the battle
+    /// </summary>
+    public void EnterBattle(ulong battleId)
+    {
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle not found.");
+
+        Assert(battle.Fee == Message.Value, "Battle fee is not matching with entry fee paid.");
+
+        var user = GetUser(battleId, Message.Sender);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        SetUser(battleId, Message.Sender, user);
+
+        var userindex = GetUserIndex(battleId);
+        Assert(userindex != MaxUsers, "Max user reached for this battle.");
+        battle.Users.SetValue(Message.Sender, userindex);
+        SetUserIndex(battleId, (userindex + 1));
+
+        SetBattle(battleId, battle);
+
+        Log(new BattleEventLog { Event = "Enter", BattleId = battleId, Address = Message.Sender });
+    }
+    /// <summary>
+    /// 4 different user will end the battle and submit the score
+    /// </summary>
+    public void EndBattle(Address userAddress, ulong battleId, uint score)
+    {
+        Assert(Message.Sender == Owner, "Only battle owner can end game.");
+
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle not found.");
+
+        var user = GetUser(battleId, userAddress);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Score = score;
+        user.ScoreSubmitted = true;
+
+        SetUser(battleId, userAddress, user);
+
+        var ScoreSubmittedCount = GetScoreSubmittedCount(battleId);
+        ScoreSubmittedCount += 1;
+        SetScoreSubmittedCount(battleId, ScoreSubmittedCount);
+
+        var highestScorer = GetHighestScorer(battleId);
+
+        if (score > highestScorer.Score)
+            SetHighestScorer(battleId, new BattleHighestScorer { Scorers = new Address[] { userAddress }, Score = score });
+        else if (score == highestScorer.Score)
+        {
+            var scorers = highestScorer.Scorers;
+            Array.Resize(ref scorers, scorers.Length + 1);
```

Comment:
This allocate new array actually and iterate it by copying fully. We should avoid it. What you need to is highest score and how many times same highest score counted and highest user. You can keep last highest one always whenever equal score submited. So you can do something like this 


```
    public struct BattleHighestScorer
    {
        public uint Score;
        public uint HighestScoreCount;
        public Address HighestScorer;
    }
```

```
      if (score > highestScorer.Score)
        {
            highestScorer.Score = score;
            highestScorer.HighestScorer = userAddress;
            highestScorer.HighestScoreCount = 1;

            SetHighestScorer(battleId, highestScorer);
        }
        else if (score == highestScorer.Score)
        {
            highestScorer.HighestScoreCount++;
            SetHighestScorer(battleId, highestScorer);
        }

        if (ScoreSubmittedCount == MaxUsers)
        {
            highestScorer = GetHighestScorer(battleId);
            if (highestScorer.HighestScoreCount > 1)
                CancelBattle(battle);
            else
                ProcessWinner(battle, highestScorer.HighestScorer);
        }
```